text/001x-battery_improvements.md

```diff
@@ -0,0 +1,294 @@
+
+  
+# Motivation
+
+The motivation is to:
+- reduce the memory required for battery reporting ([anecdotal evidence](https://github.com/ArduPilot/mavlink/pull/233#issuecomment-976197179) indicates cases with 15% of bandwidth on a default RF channel).
+- Provide a cleaner message and design for both implementers and consumers.
+- Align with DroneCan battery messages to ease transport and data conversion across different transports.
+
+The [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) has two arrays that can be used to report individual cell voltages (up to 14), or a single cumulative voltage, and which cannot be truncated.
+The vast majority of consumers are only interested in the total battery voltage, and either sum the battery cells or get the cumulative voltage.
+By separating the cell voltage reporting into a separate message the new battery status can be much smaller, and the cell voltages need only be sent if the consumer is actually interested.
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The proposed message is:
+
+```xml
+    <message id="369" name="BATTERY_STATUS_V2">
+      <description>Battery dynamic information.
+        This should be streamed (nominally at 1Hz).
+        Static/invariant battery information is sent in SMART_BATTERY_INFO.
+        Note that smart batteries should set the MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL bit to indicate that supplied capacity values are relative to a battery that is known to be full.
+        Power monitors would not set this bit, indicating that capacity_consumed is relative to drone power-on, and that other values are estimated based on the assumption that the battery was full on power-on.
+      </description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the whole battery pack (not internal electronics). INT16_MAX field not provided.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="UINT32_MAX">Battery voltage (total). UINT32_MAX: field not provided.</field>
+      <field type="int32_t" name="current" units="mA" invalid="INT32_MAX">Battery current (through all cells/loads). Positive value when discharging and negative if charging. INT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="capacity_consumed" units="mAh" invalid="UINT32_MAX">Consumed charge. UINT32_MAX: field not provided. This is either the consumption since power-on or since the battery was full, depending on the value of MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL.</field>
```

Comment:
Just working on some code that needs to use these definitions. So glad to see consumed vs remaining being separated vs in the past.  I'd like to make clearer the distinction though?


```suggestion
      <field type="uint32_t" name="capacity_consumed" units="mAh" invalid="UINT32_MAX">Consumed charge. UINT32_MAX: field not provided. This is either the consumption since power-on or since the battery was full, depending on the value of MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL. This value is monotonically increasing. Example this battery index is recharged in-flight or swapped with a full battery after landing. </field>
```

Reply:
@hendjoshsr71
1. The value depends on the flag MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL. For a smart battery you set MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL and the values capacity_consumed + capacity remaining = full battery. If that flag is not set then the consumed would be monotonically increasing - it is literal consumption.

Does that make sense?

2. Is it time for me to push the definitions into development? I think we're stabilizing re DroneCAN too right ?@dakejahl @hendjoshsr71 One more review of "the lot" would be appreciated.

Reply:
>If that flag is not set then the consumed would be monotonically increasing - it is literal consumption.

Yes this is how I see it too

>One more review of "the lot" would be appreciated.

Took another look over and LGTM

Reply:
I might add this part then to the description "For a smart battery you set MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL and the values capacity_consumed + capacity remaining = full battery. If that flag is not set then the consumed would be monotonically increasing"

I was thinking it might be the total consumption since power on which could be greater than the pack's actual capacity due to charging. Or due to refilling the fuel tank or recharging without a reboot.  (Reminds me separately it would be nice to one day for both PX4 and AP to not force other energy systems into the BATTERY_STATUS message. :)

2.) I'm good with moving forward with these so that we can take them to our respective teams. 

Probably, could have an additional comment on the meaning of `"MAV_BATTERY_STATUS_FLAGS_PROTECTIONS_ENABLED"`

Reply:
@hendjoshsr71  I don't think it would matter if the value could be recharged on the fly via the power monitor - in that case if you saw the consumed current go down you'd be fine. But you're right, battery hot-swaps are problematic. In the ideal case you'd power-cycle the power module on hot-swap. 

FMI how does the flight controller remain powered  on hot swap? Is there a backup battery on a separate power module? 

How about:

```suggestion
      <field type="uint32_t" name="capacity_consumed" units="mAh" invalid="UINT32_MAX">Consumed charge. UINT32_MAX: field not provided. If MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL is true this is the charge used since the battery was full. Otherwise it is the consumption since power-on (as measured using a power monitor). Note that if the flag is unset the value will not be useful (if you hot-swap a battery without power-cycling the power module).</field>
```

Let's finish these discussions before moving to the next stage. We're not far off.

Reply:
>In the ideal case you'd power-cycle the power module on hot-swap.

I don't think that's necessary. If you're hot swapping batteries the whole point is to keep the system turned on (think cameras on a movie set that take a long time to turn on and set up). Power modules are connected directly to the main power system (powered by 2 batteries in parallel for example), so there is no ability to reset them. If you hotswap, just be prepared to see additional consumption (2x the capacity of a single battery if you drain 2 fully charged batteries). This sounds perfectly reasonable.

>FMI how does the flight controller remain powered on hot swap? Is there a backup battery on a separate power module?

Typically the batteries are in parallel